testify/src/main/java/testify/annotation/logging/LogRecorder.java

```diff
@@ -0,0 +1,370 @@
+        this.processName = processName;
+        this.dedicatedBus = dedicatedBus;
+        ROOT_LOGGER.addHandler(handler);
+        handler.setLevel(Level.ALL);
+    }
+
+    // To test this class, it is useful to be able to retrieve some internals
+    synchronized Deque<List<LogSetting>> copySettingsStack() {
+        return settingsStack.stream()
+                .map(ArrayList::new) // copy each list
+                .collect(ArrayDeque::new, Deque::add, forbidCombining()); // collect into new deque
+    }
+
+    static String getDedicatedBusName(String processName) { return INITIAL_BUS_NAME + "#" + processName; }
+
+    // This is the only case where a lock is obtained for the recorder BEFORE the publisher.
+    // It is acceptable in this case because the recorder is not yet known to the publisher.
+    // Once initialization is complete, locks must always be acquired in publisher-then-recorder order
+    synchronized LogRecorder initialize() {
+        // prepare to receive and handle requests
+        dedicatedBus.onMsg(PUSH_SETTINGS, this::receivePushSettings);
+        dedicatedBus.onMsg(POP_SETTINGS, this::receivePopSettings);
+        dedicatedBus.onMsg(REQUEST_THREAD_TABLE, this::replyThreadTable);
+        dedicatedBus.onMsg(REQUEST_LOG_RECORDS, this::replyLogRecords);
+        if (ProcessRunner.isChildProcess()) keepAlive.begin();
+        dedicatedBus.onMsg(CLOSE, this::close);
+        // make myself known to the publisher
+        dispatchRequestAndWaitForReply(initialBus, HELLO, processName);
+        // wait for the settings
+        Deque<List<LogSetting>> stack = dedicatedBus.get(INITIALIZE_SETTINGS_STACK);
+        settingsStack.addAll(stack);
+        stream(stack.descendingIterator()).flatMap(List::stream).forEach(LogSetting::apply);
+        dispatchReply("initial settings applied");
+        return this;
+    }
+
+    static void sendInitialSettings(Bus dedicatedBus, Deque<List<LogSetting>> stack) {
+        dispatchRequestAndWaitForReply(dedicatedBus, INITIALIZE_SETTINGS_STACK, stack);
+    }
+
+    private static<K extends Enum<K>&TypeSpec<T>, T> void dispatchRequestAndWaitForReply(Bus bus, K requestType) {
+        int id = getNextRequestId(bus);
+        bus.put(requestType);
+        bus.get(String.format(RECEIPT_FORMAT, id));
+    }
+
+    private static<K extends Enum<K>&TypeSpec<T>, T> void dispatchRequestAndWaitForReply(Bus bus, K requestType, T requestPayload) {
+        final int id = getNextRequestId(bus);
+        bus.put(requestType, requestPayload);
+        bus.get(String.format(RECEIPT_FORMAT, id));
+    }
+
+    private void dispatchReply(String s) {
+        dispatchReply(dedicatedBus, s);
+    }
+
+    static void dispatchReply(Bus bus, String s) {
+        int requestId = bus.get(REQUEST_ID);
+        bus.put(String.format(RECEIPT_FORMAT, requestId), s);
+    }
+
+    private static int getNextRequestId(Bus dedicatedBus) {
+        synchronized (dedicatedBus) {
+            final int requestId;
```

Comment:
declaration and initialization of 'requestId' could be combined on one line

Reply:
Changed.